Ordered deletion of child resources

muvaf opened the pull request:
This change enables template stack developers to specify a deletion priority through the annotation `templatestacks.crossplane.io/deletion-priority`.

Note for reviewers: The first commit adds no-op finalizer add/removal logic and the second commit does a bunch of file reorganization, I would suggest you doing reviews on the last commit to see the actual changes related to ordered deletion. The rest is pretty much non-functional refactoring.

It's ready for review but I'll keep this in draft until I implement unit test for this deleter and expand the unit tests of `Reconcile` to cover deletion and finalizer logic.

Fixes #17

muvaf committed:
reconciler: add dummy add/remove finalizer logic

Signed-off-by: Muvaffak Onus <onus.muvaffak@gmail.com>

muvaf committed:
Reorganization of files with no functional change

Signed-off-by: Muvaffak Onus <onus.muvaffak@gmail.com>

muvaf committed:
Ordered deletion of child resources is implemented

Signed-off-by: Muvaffak Onus <onus.muvaffak@gmail.com>

muvaf commented:
I've manually tested this via Wordpress app by adding `templatestacks.crossplane.io/deletion-priority: "2"` to `KubernetesApplication` and `templatestacks.crossplane.io/deletion-priority: "1"` to `MySQLInstance`. As expected, the order of deletion was `KubernetesApplication` -> `MySQLInstance` -> `KubernetesCluster`. The controller won't delete the next resource until it confirms that the former priority level is gone from the cluster.

@prasek  I believe this will also help with sample stacks' problem of provider deletion; we can give `-1` priority to `Provider` resource to guarantee that it's the last one to delete.

negz commented on pkg/templating/api.go:
```diff
@@ -17,25 +17,39 @@ limitations under the License.
+	DeletionPriorityAnnotationZeroValue = "0"
+)
+
+const minInt = -(int(^uint(0) >> 1)) - 1
```
I feel like this dark magic could use a comment. It seems like it's the same as `math.MinInt64` on the Go playground - I'm guessing this is working around the fact that `strconv.Atoi` returns an `int`, which is ["either 32 or 64 bits"](https://golang.org/ref/spec#Numeric_types)? Perhaps we should use an `int64` and `strconv.ParseInt("0", 10, 8)` to bound this at 256 priority levels?

muvaf replied:
Google didn't tell me about `math.MinInt64`. Interesting. Updated with your suggestion.

negz commented on pkg/templating/api.go:
```diff
@@ -153,6 +167,72 @@ func (lo ParentLabelSetAdder) Patch(cr resource.ParentResource, list []resource.
+// deletion priority.
+func (d *APIOrderedDeleter) Delete(ctx context.Context, list []resource.ChildResource) ([]resource.ChildResource, error) {
+	hp := minInt
+	del := []resource.ChildResource{}
```
The current implementation is fine, but I'm a little curious whether it would be possible to do this using `sort.Slice`. Something like:

```go
func priority(o metav1.Object) int {
	val, ok := res.GetAnnotations()[DeletionPriorityAnnotationKey]
	if !ok {
		return 0
	}
	p, err := strconv.Atoi(val)
	if err != nil {
		return 0
	}
	return p
}

sort.Slice(list, func(i, j resource.ChildResource) bool { return priority(i) < priority(j) }
```

muvaf replied:
In one reconcile, we only operate on the highest level of priority that has some resources in the cluster, which can be extracted with one pass over the list. I thought about sorting in the beginning and then iterate over the sorted array but bookkeeping seemed similar to the existing one. But I'll give it a shot to see if the code reads better using sort.

muvaf replied:
It ended up a little more complex with the bookkeeping and there wasn't a good way to incorporate the return of error when the annotation value is not an integer, so I decided to keep it as is.

negz commented on pkg/templating/api.go:
```diff
@@ -153,6 +167,72 @@ func (lo ParentLabelSetAdder) Patch(cr resource.ParentResource, list []resource.
+		del = append(del, res)
+	}
+	for _, res := range del {
+		if err := d.kube.Delete(ctx, res); client.IgnoreNotFound(err) != nil {
```
Does this call block until the deleted resource's finalizers are gone?

muvaf replied:
No, it doesn't block. If the resource still exists, that priority level would be still the highest one in the next reconcile.

hasheddan replied:
If we wanted it to block I believe we could do foreground deletion here. I guess the tradeoff would be that we might block until reconcile times out, but we would also reduce the number of reconciles because we wouldn't keep trying to delete every second if it was taking a resource a little while to get cleaned up. In the case of reconciliation time out we would be requeued anyway. In practice, the behavior to the end user would look the same I believe.

muvaf replied:
I'd not like to block the process in the reconcile and timeout in an expected scenario. The processing of event queue would be slower as there are other objects and this reconciler doesn't utilize concurrency for processing the queue of CRs.

negz commented on pkg/templating/reconciler.go:
```diff
@@ -198,6 +199,27 @@ func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	}
 
+	if meta.WasDeleted(cr) {
+		deleting, err := r.children.Delete(ctx, childResources)
```
Is there any value in confirming that we're the controller reference of the children we're about to delete before we do so, now that we're doing this manually?

muvaf replied:
It's a rare case but it'd be consistent if we do that both for deletion as well as during `Apply`. I'll add it.

muvaf reviewed (commented):
@negz thanks for the review. I'll address the concerns mentioned in my responses tomorrow and ping you for final review.

muvaf committed:
Use math.MinInt64 as minimum value of deletion priority order

Signed-off-by: Muvaffak Onus <onus.muvaffak@gmail.com>

muvaf committed:
Implement unit tests for ordered child deleter

Signed-off-by: Muvaffak Onus <onus.muvaffak@gmail.com>

muvaf committed:
Implement additional unit tests for reconciler to test the deletion logic

Signed-off-by: Muvaffak Onus <onus.muvaffak@gmail.com>

muvaf committed:
add controller checks for apply and delete functions

Signed-off-by: Muvaffak Onus <onus.muvaffak@gmail.com>

muvaf committed:
Remove provider exception in owner reference adder since we can tune its deletion priority now

Signed-off-by: Muvaffak Onus <onus.muvaffak@gmail.com>

muvaf commented:
@negz Addressed your comments. Tested this with app-wordpress (giving KA high priority to be deleted first) and sample-stack-gcp (giving provider negative priority to be deleted last) and confirmed all is working as intended.

negz commented:
Thanks @muvaf! I think this is quite a simple and pragmatic solution to the issues we've faced around deletion order.

prasek commented:
> I've manually tested this via Wordpress app by adding `templatestacks.crossplane.io/deletion-priority: "2"` to `KubernetesApplication` and `templatestacks.crossplane.io/deletion-priority: "1"` to `MySQLInstance`. As expected, the order of deletion was `KubernetesApplication` -> `MySQLInstance` -> `KubernetesCluster`. The controller won't delete the next resource until it confirms that the former priority level is gone from the cluster.
> 
> @prasek I believe this will also help with sample stacks' problem of provider deletion; we can give `-1` priority to `Provider` resource to guarantee that it's the last one to delete.

@muvaf awesome, thanks! Can you drop your changes into app-wordpress as part of the v0.11 release? Would be great to test it out in the next day or two.

prasek commented:
actually @muvaf I'll submit an app-wordpress PR, seems straightforward. thx.
